### PR TITLE
ArC - improve error message when annotation default value cannot be read

### LINKED ISSUE #1832: NPE from ArcAnnotationProcessor when default value() is used
Hello @mkouba ,

This is in regards to the issue we discussed today on Zulip.

After using `jandex-maven-plugin` to generate `META-INF/jandex.idx` I'm getting errors from all my tests.

Maven plugin I used:

```xml
<plugin>
  <groupId>org.jboss.jandex</groupId>
  <artifactId>jandex-maven-plugin</artifactId>
  <version>1.0.5</version>
  <executions>
    <execution>
      <id>make-index</id>
      <goals>
        <goal>jandex</goal>
      </goals>
    </execution>
  </executions>
</plugin>
```

And the stack trace I'm getting is:

```plain
Caused by: java.lang.NullPointerException: Value not set for java.lang.Class<? extends akka.actor.Actor> value()
    at io.quarkus.arc.processor.AnnotationLiteralProcessor.process(AnnotationLiteralProcessor.java:92)
    at io.quarkus.arc.processor.BeanGenerator.initConstructor(BeanGenerator.java:624)
    at io.quarkus.arc.processor.BeanGenerator.createConstructor(BeanGenerator.java:513)
    at io.quarkus.arc.processor.BeanGenerator.generateProducerMethodBean(BeanGenerator.java:371)
    at io.quarkus.arc.processor.BeanGenerator.generate(BeanGenerator.java:124)
    at io.quarkus.arc.processor.BeanProcessor.process(BeanProcessor.java:186)
    at io.quarkus.arc.deployment.ArcAnnotationProcessor.build(ArcAnnotationProcessor.java:259)
```

The source code (at revision with jandex failure):

https://github.com/sarxos/abberwoult/tree/00dd5d0a182f8ea85657044a4e4a228a7d828756

The problematic annotation is this buddy here:

```java
@Qualifier
@Documented
@Target({ FIELD, PARAMETER, METHOD })
@Retention(RUNTIME)
public @interface ActorByClass {

	final Class<? extends Actor> NO_CLASS = Actor.class;

	@Nonbinding
	Class<? extends Actor> value() default Actor.class;
}
```

After you suggested to override jandex dependency in corresponding maven plugin to 2.1.1.Final all tests are again passing, which proves that `jandex-maven-plugin` uses outdated dependency.

The workarounds so far:

1) Use empty `META-INF/beans.xml` instead of `jandex-maven-plugin`, or
2) Force newer (2.1.1.Final) jandex dependency in `jandex-maven-plugin`:

```xml
<plugin>
  <groupId>org.jboss.jandex</groupId>
  <artifactId>jandex-maven-plugin</artifactId>
  <version>1.0.5</version>
  <executions>
    <execution>
      <id>make-index</id>
      <goals>
        <goal>jandex</goal>
      </goals>
    </execution>
  </executions>
  <dependencies>
    <dependency>
      <groupId>org.jboss</groupId>
      <artifactId>jandex</artifactId>
      <version>2.1.1.Final</version>
    </dependency>
  </dependencies>
</plugin>
```

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #1832

### CODE REVIEW

**Comment by mkouba** [comment]
@sarxos I hope the updated error message is a little bit more clear ;-)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationLiteralGenerator.java [hunk 1]
```diff
@@ -142,45 +137,12 @@ static void createAnnotationLiteral(ClassOutput classOutput, ClassInfo annotatio
...
-                        }
-                        break;
-                    default:
-                        break;
-                }
-            } else {
-                retValue = loadValue(valueMethod, value, annotationClass, method);
+                throw new NullPointerException(String.format(
+                        "Value is not set for %s.%s(). Most probably an older version of Jandex was used to index an application dependency. Make sure that Jandex 2.1+ is used.",
+                        method.declaringClass().name(), method.name()));
```
Comment (gsmet) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationLiteralGenerator.java]: So we don't need to handle the primitive case anymore?

**Comment by sarxos** [comment]
@mkouba It's ok, however @dmlloyd suggested to change `NullPointerException` to some other exception designed specifically for processing failures. See our discussion in _annotations_ topic on Zulip.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationLiteralGenerator.java (hunk 1 above)
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationLiteralGenerator.java]: We do handle primitives in `AnnotationLiteralGenerator.loadValue()`. I've just removed the part which sets the values to the default JLS values - this wouldn't help much and the produced literal would be incorrect.

**Comment by mkouba** [comment]
@sarxos Could pls you verify the re-indexing fallback works?

**Comment by sarxos** [comment]
Sure, let me checkout this PR.

**Comment by sarxos** [comment]
@mkouba 

I confirm. Re-indexing fallback works as expected. Before your patchset has been applied all my tests were failing. After patchset was applied all my tests passed. I also observed warning you included in a `handleFilePath(Path)` method of `ApplicationArchiveBuildStep` class:

```plain
17:52:20.415 [build-4] DEBUG i.q.d.i.ApplicationArchiveBuildStep - Indexing dependency: /home/sarxos/.m2/repository/io/smallrye/smallrye-config/1.3.5/smallrye-config-1.3.5.jar
17:52:20.429 [build-4] DEBUG i.q.d.i.ApplicationArchiveBuildStep - Indexing dependency: /home/sarxos/workspace-2019/abberwoult/abberwoult-core/target/test-classes
17:52:20.439 [build-4] DEBUG i.q.d.i.ApplicationArchiveBuildStep - Indexing dependency: /home/sarxos/workspace-2019/abberwoult/abberwoult-annotations/target/abberwoult-annotations-0.0.1-SNAPSHOT.jar
17:52:20.493 [build-4] WARN  i.q.d.i.ApplicationArchiveBuildStep - Re-indexing /home/sarxos/workspace-2019/abberwoult/abberwoult-annotations/target/abberwoult-annotations-0.0.1-SNAPSHOT.jar - at least Jandex 2.1 must be used to index an application dependency
17:52:20.509 [build-8] DEBUG io.quarkus.deployment.index - Index bean class: io.smallrye.config.inject.ConfigProducer
17:52:20.511 [build-8] DEBUG io.quarkus.deployment.index - Index annotation: javax.enterprise.inject.Produces
17:52:20.511 [build-8] DEBUG io.quarkus.deployment.index - Index annotation: javax.enterprise.context.Dependent
```

:+1:

**Review by sarxos — APPROVED** [review]
Fixes indexing issue. :+1:

**Comment by mkouba** [comment]
@sarxos Thanks for verification ;-)

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Merged, thanks!

**Pull request merged**